src/autoschedulers/adams2019/State.h

```diff
@@ -0,0 +1,79 @@
+#ifndef STATE_H
+#define STATE_H
+
+#include "ASLog.h"
+#include "CostModel.h"
+#include "DefaultCostModel.h"
+#include "Featurization.h"
+#include "Halide.h"
+#include "LoopNest.h"
+#include "PerfectHashMap.h"
+
+namespace Halide {
+namespace Internal {
+namespace Autoscheduler {
+
+struct State {
+    mutable RefCount ref_count;
+    IntrusivePtr<const LoopNest> root;
+    IntrusivePtr<const State> parent;
+    double cost = 0;
+    int num_decisions_made = 0;
+    bool penalized = false;
+    string schedule_source;
+
+    static int cost_calculations;
+
+    State() = default;
+    State(const State &) = delete;
+    State(State &&) = delete;
+    void operator=(const State &) = delete;
+    void operator=(State &&) = delete;
+
+    uint64_t structural_hash(int depth) const;
+
+    // Compute the parent and depth of every loop nest node
+    void compute_loop_nest_parents(map<const LoopNest *, pair<const LoopNest *, int>> &p,
```

Comment:
Now that this is all an API exposed in a header file it could use comments explaining what each method does here. E.g. it's not clear to me why deepest_common_ancestor is a member function at all, given that it takes two loop nests.

Reply:
You're completely right, it looks like `deepest_common_ancestor` and `compute_loop_nest_parents` have no reason to be member functions of `State`. I removed these and put them into LoopNest.h/LoopNest.cpp in c099ebb